### ISSUE TITLE
Native Downloads

### ISSUE THREAD

**Issue by jaitaiwan** [tcz/PHPTracker#43]
### About
Currently the PHPTracker project doesn't have the ability to download a torrent which in reality is actually a crucial part of the BitTorrent eco system. From the specification, it appears that it groups the tasks of seeding and downloading into the function of a client. So having this missing from the library is quite paramount in its adoption.

### Rational
The reason for implementing this in v0.4 is that it is a crucial part of the BitTorrent eco-system and personally a required feature to my [Provender](https://bitbucket.org/centricwebestate/provender) project getting up and running.

### Purpose
This issue will track the discussion around adding this feature as well as its progress and implementation.

### Implementation Details
This will likely be as simple as adding what appears to be missing portions of code into the lib/PHPTracker/Seeder/peer.php file as well as any appropriate tests.

**Comment by jaitaiwan** [comment]
Looks like this might be a lot more work than I originally thought. It appears as if some things are hardcoded to happen such as telling a client that a seeder has all the pieces which in the case of a downloading client might not be entirely true at all.

@tcz I know your awol atm but any chance you can weigh in on how much work you think this will take? I get the feeling it's almost a complete rewrite...

**Comment by genaside** [comment]
I was sort of thinking about this too. Actualy the thing is I want my own main server to seed while the client servers do both(seeding+downloading). i tried many ideas, but not one is any good practically. I couldn't figure out why other php torrent downloaders use 3rd party programs like transmission as their downloader and not native php, isn't php up for it? If it is, then what do think about the creation of a libtorrent for php? meh too much work :)... actually how beneficial would it be to have a php version of libtorrent? Anyways, ill try playing in peer.php, and report back if i have any ideas.

**Comment by jaitaiwan** [comment]
@genaside I'm pretty much in the same position. I'm having to not use BitTorrent to do code deployments because there's practically no libraries which support downloading as well as seeding.

I've been trying to hack away at Peer.php but I'm honestly finding that a lot of the code around peers, announce trackers and process forking two intertwined.

@tcz are you coming back?

**Comment by genaside** [comment]
Ok, i plan to work on it this weekend. What do you think about me spliting up the program, one phptracker and one phpclient? I can probably use phptracker and sort of morph it into a client. I notice in some programs with a tracker and a torrent they have different numbers. Two seperate programs can work as one depending what the admin wants. I dont know. What do you think about all this?

**Comment by jaitaiwan** [comment]
From what I see of the Torrent spec, a seeder and downloader are
practically just defining two roles of the client. Meaning that we should
build a client which seeds AND downloads. I'm not sure though if its
possible to split the two functions out but thatd be your call if you're
working on it.

Personally I feel the posix forking stuff should be moved out from the
client entirely first before adding new functionality. But that's not my
call and I'm waiting for @tcz to get back from wherever he is to confirm.

Also in the meantime you should be able to use #42 (if I remember the pull
request number correctly) to base your branch off so that you can get
passing tests to work from.

**Comment by genaside** [comment]
I manage to create a test function in core.php that reads the torrent file and gets the announcer addresses, then i announce myself to the tracker, get the peer-list, make a handshake with one of the peers, then i sent an unchoke message. Doing all that is one thing making a suitable client is another.

Some thoughts:
* A ubuntu tracker said my client(phptracker) is too old, and i need upgrading. Not sure what is meant by this, probably tracker's white-list for torrent clients. I used "-PT0031-" for the client id btw.
* I have no idea to connect to trackers like udp://open.demonii.com:1337/announce, I think i did connect to them but they're not returning a response. I can connect to the udp and host, but what is with path all about? Would UDP tracker be a good feature to add to PHPTracker?
* Forker gets the job done, but it irks me.(i feel like i should move to python and threads). I don't know.

**Comment by jaitaiwan** [comment]
Good Job, quicker than I am for sure.

I don't think PHPTracker supports UDP which might be why it thinks the client is too old... Also addresses your second point.

Forker does get the job done, but it's something that should be separated from the client itself. Python would probably defeat the purpose of having an all-php tracker though considering you can just use LibTorrent with python which is stable.

**Comment by jaitaiwan** [comment]
The idea of splitting the forking out of the mix is because you could use PHPTracker with another daemonizing program.

**Comment by genaside** [comment]
Sorry for the long wait, been busy with school and all. Ok, I notice more trackers are not supporting my client. I think there might be a place to register -PT0031- or -PT0040- , but idk. If you use the client with other PHPtrackers then it works beautifully. Right now iam try to improve and add to it. Is there anything else you want the client todo? something simple though :)
P.S 
I notice tcz maked issue as ready, what does that mean, did he already create the native download feature?

**Comment by jaitaiwan** [comment]
Not even sure tbh. I don't know why @tcz's been silent.

**Comment by jaitaiwan** [comment]
@genaside got skype?

**Comment by genaside** [comment]
@jaitaiwan 
No, and no facebook either. my paranoia prevents me from tying my face to my work. However iam on freenode(irc) under the same name.

**Comment by jaitaiwan** [comment]
Sweet, I'll catch up with you there.

**Comment by genaside** [comment]
During testing I notice i couldn't connect to 80 percent of peers. I went and check their their addres:port with another program. Research points to a thing "Nat punchthough", it sort of makes sense why with other clients iam able to connect with other peers without opening ports on the router.
So does anyone know how would implement Nat punch though for php?

**Comment by jaitaiwan** [comment]
I'll do some research and find out

**Comment by genaside** [comment]
Ok update. this is what I have completed so for on the daemon(native downloader).
* Added support for connecting and receiving messages from a UDP trackers.
* Even though I am still thinking about supporting magnet links, users can still grab .torrrent data from a local source or remote URL.
* Multiple file support for torrents is now supported.
* Since the daemon can't be control directly, a socket is used to control the behaviour of the daemon.
You can add torrents, remove torrents, view torrent status,  or just exit the daemon.
* Since the socket can be accessed remotely I added a primitive authentication(user pass) to make sure not just anybody can mess with the torrents.

Ok, this is a new program, even though it's still unfinished and has many bugs(and grammer/spelling issues), the program works. I pass the location(url) of the torrent, path to a local destination,  and set active option to true the torrent begins downloading. Torrent uploads too,

The problem. When the native downloader was started I thought it will be bundle with PHPTracker, but I have second thoughts because it's a separate program. What do you think? PHPTracker should be bundle with an UDP version and HTTP version, but I think some people will want a tracker and others will want a tracker rarely people want both. I don't know, tell me what you guys think. 

Btw a UDP Tracker is better than a HTTP tracker, well I guess they both have pros and cons. A UDP tracker can sit on your machine like a standalone sever, while HTTP tracker starts from apache .

And one more things, maybe tonight or tomorrow I will upload the code to my git-hub profile and stop working on it until I finish creating a useful library that will PHPTorrent and PHPTracker's bit array and 64bit intergers

**Comment by jaitaiwan** [comment]
Did you ever figure out why PHPTracker was apparently an "old client"?

**Comment by genaside** [comment]
No idea, however I notice that i was getting  that error from trackers ubuntu.com  and gentoo.org, but since then Ive been testing it on opentorrent, naay torrent, and a whole bunch of udp trackers, didnt see that error anymore.

**Comment by darksom** [comment]
is phptracker  dead?

**Comment by jaitaiwan** [comment]
Tcz just doesn’t have time to do stuff to it. We’ve been working on LibPHPTorrent for some time now

**Comment by tcz** [comment]
Yes, pretty much dead from my end but contributions are welcome.
